plugin/pkg/kmsproviders/vault/vault.go

```diff
@@ -0,0 +1,240 @@
+		}
+		count++
+	}
+
+	if config.RoleID != "" || config.SecretID != "" {
+		if config.RoleID == "" {
+			return errors.New("vault provider has invalid approle authentication information")
+		}
+		count++
+	}
+
+	if count == 0 {
+		return errors.New("vault provider has no authentication information")
+	}
+	if count > 1 {
+		return errors.New("vault provider has more than one authentication information")
+	}
+
+	return nil
+}
+
+type vaultEnvelopeService struct {
+	config *EnvelopeConfig
+	client *clientWrapper
+	// We may update token for api.Client, but there is no sync for api.Client.
+	// Read lock for encrypt/decrypt requests, write lock for login requests which
+	// will update token for api.Client.
+	rwmutex sync.RWMutex
+}
+
+func (s *vaultEnvelopeService) Decrypt(data string) ([]byte, error) {
+	// Find the mached key
+	var key string
+	for _, name := range s.config.KeyNames {
+		if strings.HasPrefix(data, name+":") {
+			key = name
+			break
+		}
+	}
+	if key == "" {
+		return nil, errors.New("no matching vault key found")
+	}
+
+	// Replace the key name with "vault:" for Vault transit API
+	if !strings.HasPrefix(data, key) {
+		return nil, errors.New("encrypted data from storage does not have key prefix")
+	}
+	cipher := vaultPrefix + data[len(key):]
+
+	plain, err := s.withRefreshToken(false, key, cipher)
+	if err != nil {
+		return nil, err
+	}
+
+	return base64.StdEncoding.DecodeString(plain)
+}
+
+func (s *vaultEnvelopeService) Encrypt(data []byte) (string, error) {
+	// Use the frist key to encrypt
+	key := s.config.KeyNames[0]
+	plain := base64.StdEncoding.EncodeToString(data)
+
+	cipher, err := s.withRefreshToken(true, key, plain)
+	if err != nil {
+		return "", err
+	}
+
+	// The format of cipher from Vault is "vault:v1:....".
+	// "vault:" is unnecessary, replace it with key name.
+	if !strings.HasPrefix(cipher, vaultPrefix) {
+		return "", fmt.Errorf("encrypted data from vault does not have prefix %v", vaultPrefix)
+	}
+	return key + cipher[len(vaultPrefix):], nil
+}
+
+func (s *vaultEnvelopeService) withRefreshToken(isEncrypt bool, key, data string) (string, error) {
+	// Execute operation first time.
+	var result string
+	var err error
+	func() {
+		s.rwmutex.RLock()
+		defer s.rwmutex.RUnlock()
+		if isEncrypt {
+			result, err = s.client.encrypt(key, data)
+		} else {
+			result, err = s.client.decrypt(key, data)
+		}
+	}()
+	if err == nil || s.config.Token != "" {
+		return result, err
+	}
+	_, ok := err.(*forbiddenError)
+	if !ok {
+		return result, err
+	}
+
+	// The request is forbidden, refresh token and execute operation again.
+	// With the expected usage:
+	//a. rare calls to KMS provider for decrypt during secret read: due to DEK caching,
+	//b. rare concurrent secret creation,
+	//c. Token policy having reasonable expiry and num. of use;
+	// the locking is not expected to degrade performance for normal usage scenarios.
+	// The race condition is still eliminated for worst case scenarios.
+	s.rwmutex.Lock()
```

Comment:
@ericchiang I think this was the area you were commenting on before.  Looks like the diff changed.

Reply:
This lock is still held when performing the HTTP request to vault. I noted that it only needs to be held while rotating/accessing the bearer token.

It's not a deal breaker, but holding mutex's while performing HTTP requests has lead to some _really_ bad behavior in the OpenID Connect plugin code (like crashing the API server).

Reply:
> This lock is still held when performing the HTTP request to vault. I noted that it only needs to be held while rotating/accessing the bearer token.
> 
> It's not a deal breaker, but holding mutex's while performing HTTP requests has lead to some really bad behavior in the OpenID Connect plugin code (like crashing the API server).

Thought that was the concern.

@vineet-garg this should be an easy fix.  Use the same anonymous func approach and only hold the write lock as long as is absolutely necessary.  Then acquire the read lock for the rest of the function.

Reply:
@deads2k The locks are already acquired only for the minimum absolutely necessary duration necessary to implement the logic.  I am not able to find any way of further optimization. Please follow my last response on the conversation yesterday, I have explained what is the current logic and why it is necessary to lock. I am open to suggestions.

Reply:
> @deads2k The locks are already acquired only for the minimum absolutely necessary duration necessary to implement the logic. I am not able to find any way of further optimization. Please follow my last response on the conversation yesterday, I have explained what is the current logic and why it is necessary to lock. I am open to suggestions.

read locks and write locks are different and different exclusion rules. 

```go
func(){
  s.rwmutex.lock
  defer s.rwmutex.unlock
  // double check the status
  s.client.refreshToken
}

s.rwmutex.rlock
defer s.rw.mutext.runlock
if isEncrypt{} else{}
return
```

Reply:
Just pasted an example of how to do it.  You just have to find a way to make the double-check cheap.

Reply:
@deads2k as I explained earlier, the second encrypt/decrypt request has to be made while holding the mutual exclusive lock along with refreshToken.

Reply:
> @deads2k as I explained earlier, the second encrypt/decrypt request has to be made while holding the mutual exclusive lock along with refreshToken.

Just setting and getting the token via an accessor which clears bad/dead tokens with a recursive all to this method would clear it.

I think its an easy problem to solve.  I don't really care if you solve it here, but you will regret not solving it.

Reply:
As stated above, I think this can be easily solved by being able to pass a token to encrypt/decrypt

```
s.mu.Lock()
token := s.currentToken
s.mu.unlock()

err := s.encrypt(token, data)

if !needToRotate(err) {
    return
}

s.mu.Lock()
if token == s.currentToken {
    // token hasn't been rotated, rotate it.
    s.currentToken = s.refreshToken()
}
token = s.currentToken
s.mu.Unlock()

err := s.encrypt(token, data)
```

The second request will always use a new token and you're not at risk of two processes rotating the same token twice.

Reply:
@ericchiang I do get what you are suggesting. 
My earlier response went into the outdated conversation, I am putting it here again. Please let know if it makes sense. I can elaborate it more if the explanation is not clear.

Every thread that fails with 403 needs to rotate the token separately, One failed thread cannot rely on the token that was rotated by the previous thread. The token rotated by a different thread could have got expired, exhausted its number of use or had been revoked explicitly by the time another thread gets to execute.
so "if token == s.currentToken" goes not guarentee to work in the worst case.
Also "s.mu.Unlock()" after refresh token does not guarantee to work because it is the 2nd and last retry of encrypt/decrypt and it MUST pass, The thread that rotated the token should be the one that consumes it before other threads exhaust it.

We started with exactly same code what you have in the suggestion but changed it after the issues were surfaced.

The Logic is as follows

    A number (say n) of encrypt and decrypt HTTP calls to KMS provider will run concurrently at a given time.
    Due to token time expiration or token number of use expiration or due to token revocation, one request will fail with 403. This thread will wait for other concurrent HTTP calls to finish. Other concurrent calls will also fail with 403.
    The first failed thread that acquired exclusive lock will make a token refresh HTTP call and repeat the encrypt or decrypt HTTP call that originally failed and release the lock.
    other (n-1) threads will do the same thing as in 3.
    Once all n threads (plus any new thread that got started in between) that failed pass, the concurrent execution will resume again.
    a. This logic needs HTTP requests to be blocked
    b. The assumption is n is low for reasons stated in code comments.
    c. by good use of token 'time to live' and 'number of use policy', 403 situations can be made less frequent. Explicit revocation of token will incurr 403, but it is not expected to be often.
    d. To further avoid the 403 situations, we planned to add (in a separate follow-on PR) a token renew feature which, if configured, will proactively renew the same token before its expiration. This Token renew HTTP request can be made concurrently with encrypt and decrypt HTTP requests.